test/src/596-catalogbalancer/main

```diff
@@ -0,0 +1,156 @@
+cvmfs_test_name="Catalog balancer"
+cvmfs_test_autofs_on_startup=false
+
+add_files() {
+  local repo_dir=$1
+  local num_files=$2
+  mkdir -p "$repo_dir"
+  local body="$repo_dir/file"
+
+  for i in $(seq 1 1 $num_files); do
+    touch "$body$i"
+  done
+}
+
+add_catalog() {
+  local repo_dir=$1
+  touch "$repo_dir/.cvmfscatalog"
+}
+
+cvmfs_run_test() {
+  apache_switch on
+  local repo="test.local"
+  local repo_dir="/cvmfs/$repo"
+
+  create_repo "$repo" $(whoami)
+
+  sudo bash -c "echo 'CVMFS_AUTOCATALOGS=true
+  CVMFS_AUTOCATALOGS_MAX_WEIGHT=20
+  CVMFS_AUTOCATALOGS_MIN_WEIGHT=8' >> /etc/cvmfs/repositories.d/$repo/server.conf"
+
+  start_transaction "$repo"
+  add_files "$repo_dir" 25
+  add_files "$repo_dir/dir100" 100
+  add_files "$repo_dir/dir3" 3
+  add_files "$repo_dir/dir4" 4
+  add_catalog "$repo_dir/dir4"
+  publish_repo "$repo"
+
+  echo "Checking first iteration"
+  ! [ -f "$repo_dir/dir3/.cvmfscatalog" ]                     || return 1
+  ! [ -f "$repo_dir/dir3/.cvmfsautocatalog" ]                 || return 2
+  if check_catalog_presence "/dir3" "$repo"; then return 101; fi
+  ! [ -f "$repo_dir/.cvmfscatalog" ]                          || return 3
+  ! [ -f "$repo_dir/.cvmfsautocatalog" ]                      || return 4
+  if ! check_catalog_presence "/" "$repo"; then return 101; fi
+    [ -f "$repo_dir/dir100/.cvmfscatalog" ]                   || return 5
+    [ -f "$repo_dir/dir100/.cvmfsautocatalog" ]               || return 6
+  if ! check_catalog_presence "/dir100" "$repo"; then return 101; fi
+    [ -f "$repo_dir/dir4/.cvmfscatalog" ]                     || return 7
+  ! [ -f "$repo_dir/dir4/.cvmfsautocatalog" ]                 || return 8
+  if ! check_catalog_presence "/dir4" "$repo"; then return 101; fi
+
+  start_transaction "$repo"
+  add_files "$repo_dir/dir/dir30" 30
+  for i in $(ls "$repo_dir" | grep "file"); do
+    rm "$repo_dir/$i"
+  done
+  publish_repo "$repo"
+
+  echo "Checking second iteration"
+  ! [ -f "$repo_dir/dir3/.cvmfscatalog" ]                     || return 1
```

Comment:
I think there is a limit at 255.  The point is, if possible the return code should uniquely identify the failure location.